Add PSA to block host field in probe/lifecycle handlers (#4942)
kubernetes/enhancements: keps/sig-auth/4940-psa-block-host-field-in-probes/README.md

- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

We have a `Host` field that can be set from `TCPSocketAction` and
`HTTPGetAction` fields which are part of the `ProbeHandler` and

> `LifecycleHandler` structs in Kubernetes. The `Host` field is used

Reviewer: these structs are used in the following spots in the pod spec:

```
staging/src/k8s.io/api/core/v1/types.go

container / initContainer fields:
  2872,17: 	LivenessProbe *Probe `json:"livenessProbe,omitempty" protobuf:"bytes,10,opt,name=livenessProbe"`
  2878,18: 	ReadinessProbe *Probe `json:"readinessProbe,omitempty" protobuf:"bytes,11,opt,name=readinessProbe"`
  2887,16: 	StartupProbe *Probe `json:"startupProbe,omitempty" protobuf:"bytes,22,opt,name=startupProbe"`
  2891,13: 	Lifecycle *Lifecycle `json:"lifecycle,omitempty" protobuf:"bytes,12,opt,name=lifecycle"`

(ephemeral container copies)
  4879,17: 	LivenessProbe *Probe `json:"livenessProbe,omitempty" protobuf:"bytes,10,opt,name=livenessProbe"`
  4882,18: 	ReadinessProbe *Probe `json:"readinessProbe,omitempty" protobuf:"bytes,11,opt,name=readinessProbe"`
  4885,16: 	StartupProbe *Probe `json:"startupProbe,omitempty" protobuf:"bytes,22,opt,name=startupProbe"`
  4888,13: 	Lifecycle *Lifecycle `json:"lifecycle,omitempty" protobuf:"bytes,12,opt,name=lifecycle"`
```

I'm assuming we will enforce all of those consistently, both for initContainer / container / ephemeral containers.

Author: yes correct, I will call this out clearly in this part.

Author: I think init and containers are ok but ephemeral doesn't support probes it looks like

for allowing users to specify another entity other than the
podIP (which is the default value) to which Kubelet should
perform probes to. However this opens it up for security attacks
since the `Host` field can be set to pretty much any value in the
system including security sensitive external hosts or localhost on
the node. Kubelet will be probing this set `Host` value which can
lead to blind SSRF attacks.

### Goals

> 

Reviewer: also update the baseline pod security standard at https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline for this (PSA is one implementation of PSS)

Author: ack done.

* Add  Pod Security Admission (PSA) to enable admins to restrict
  users from creating probes with the `Host` field set.
* Deprecate `.Host` field from the API by updating the types.go
  doc string and any documentation updates on the website.

### Non-Goals

> N/A

Author: yup thanks @BenTheElder ! updated this to reflect that.

that does exactly that. Once we add the PSA, we can also deprecate the API
as part of this KEP itself.

### Risks and Mitigations

There might be users who depend on the `Host` field in
their existing probes which will continue to work and if
newly created probes also need the `Host` field to point
to an external destination then the admin can avoid enforcing
the PSA to block it.

## Design Details

Add a Baseline APILevel Pod Security Admission policy to allow admins of the
cluster to block users from setting `.host` field in:

* container.LivenessProbe.ProbeHandler.HTTPGet.Host
* container.ReadinessProbe.ProbeHandler.HTTPGet.Host
* container.StartupProbe.ProbeHandler.HTTPGet.Host
* container.LivenessProbe.ProbeHandler.TCPSocket.Host
* container.ReadinessProbe.ProbeHandler.TCPSocket.Host
* container.StartupProbe.ProbeHandler.TCPSocket.Host
* container.Lifecycle.PostStart.TCPSocket.Host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* container.Lifecycle.PreStop.TCPSocket.Host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* container.Lifecycle.PostStart.HTTPGet.Host

> * container.Lifecycle.PreStop.HTTPGet.Host

Reviewer: `container` is not a full field path... explicitly specify initContainer, container, ephemeralContainers

see other examples at https://kubernetes.io/docs/concepts/security/pod-security-standards/#baseline

```
spec.containers[*]...
spec.initContainers[*]...
spec.ephemeralContainers[*]...
...
```

Author: I've updated this except the ephemeralcontainers part cause i think probes with ephemeral are not supported according to the types.go file

None

##### Unit tests

Necessary unit tests will be added to the [PSA package] for
testing the new code.
Current test coverage status for the package is:
- `k8s.io/pod-security-admission/policy`: `2025-05-06` - `89.9%`
- `k8s.io/pod-security-admission/test`: `TBD` - `TBD`

[PSA package]: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/pod-security-admission/policy

##### Integration tests

The following integration tests will be added to verify the PSA validation logic:

1. Test that pods with `.host` field set in probes are rejected when PSA is enabled with baseline level
2. Test that pods without `.host` field set in probes are allowed when PSA is enabled with baseline level
3. Test that existing pods with `.host` field set continue to work when PSA is enabled
4. Test that pods with `.host` field set are allowed when PSA is disabled or using an older version

These tests will be added to:
- `test/integration/auth/podsecurity_test.go`
https://storage.googleapis.com/k8s-triage/index.html?test=TestPodSecurity

The integration tests will verify the PSA policy validation logic by:
- Creating test cases for each probe type (HTTPGet, TCPSocket)

> - Testing each probe location (LivenessProbe, ReadinessProbe, StartupProbe, Lifecycle)

Reviewer: make sure there's a test exercising PSA validation of ephemeral containers when added via the `update pods/ephemeralcontainers` subresource

Author: goodpoint! added this also to the test case. I wasn't aware day2 pod can be modified like that. thanks for pointing this out!

Author: @liggitt I had almost committed this when I noticed:
```
	// Probes are not allowed for ephemeral containers.
	// +optional
	LivenessProbe *Probe
	// Probes are not allowed for ephemeral containers.
	// +optional
	ReadinessProbe *Probe
	// Probes are not allowed for ephemeral containers.
	// +optional
	StartupProbe *Probe
	// Lifecycle is not allowed for ephemeral containers.
	// +optional
	Lifecycle *Lifecycle
```
so I assume these fields are not applicable to ephemeral containers?

- Verifying the PSA policy enforcement at the baseline level
- Testing the behavior with different PSA configurations

##### e2e tests

There are no Pod Security specific E2E tests (we rely on integration test coverage instead), but the Pod Security admission controller is enabled in E2E clusters, and all E2E test namespaces are labeled with the enforcement label for Pod Security.

### Graduation Criteria

The PSA added will be done within a single release
and given there will be no feature gates for that,
there is no need for multi-release graduation criteria.
All related code will land within the same single release

> #### Deprecation

Reviewer 2: So they're deploying the apiserver as a host-network pod and specifying to use `host: 127.0.0.1` for the probes, presumably because there may be a firewall blocking access to the healthz port on the public node IP. That seems reasonable. (Ah, actually, there are examples of this in k/k/cluster/ too.)

We could have the semantics of the deprecation/PSA be "you aren't allowed to specify `host` except in the case of specifying `127.0.0.1` or `::1` for a `hostNetwork: true` pod...

> #### Deprecation

Author: ```
We could have the semantics of the deprecation/PSA be "you aren't allowed to specify host except in the case of specifying 127.0.0.1 or ::1 for a hostNetwork: true pod...
```
hmm I could but its still a bit mucky, i.e we either deprecate the field or we don't right? Is it OK to deprecate with exceptions :D

> #### Deprecation

Reviewer 3: I'm honestly not sure what the most reasonable option is here. cc @aojea @thockin @liggitt

> `.Host` field will be deprecated from the Probe and Lifecycle Handlers

Reviewer 2: Hm... maybe remove the `Deprecation` heading and just include this as part of `Graduation Criteria`.

Since the deprecation consists _only_ of marking the field deprecated, and there is no intention of removing it or introducing any replacement for it, this is also something that can happen in a single release.

### Upgrade / Downgrade Strategy

Any older pods with this field set should not be affected
with the above solution. Only newer pods getting created
with the field will be alerted.

Users who are using this field can switch to using exec
probes moving forward which should unblock them given exec
probes can provide the same functionality.

### Version Skew Strategy

N/A since its only within a single component: pod-security-admission
and doesn't cross multiple components.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

We decided to not go with feature gates and use PSA versioning.

> So if the admin sets pod-security.kubernetes.io/enforce-version: v1.33

Reviewer: ```suggestion
So if the admin sets pod-security.kubernetes.io/enforce-version: v1.34
```

Author: thanks, updated!

We decided to not go with feature gates and use PSA versioning.
So if the admin sets pod-security.kubernetes.io/enforce-version: v1.34
on a namespace this feature will get enabled.

###### Does enabling the feature change any default behavior?

* There is no effect on clusters where PSA is not enabled OR an older
PSA version is used.

* There is no effect on clusters where `.Host` probes are not used

* There is no effect on clusters where an older PSA versioning is being
used

* If users create pod probes with `.Host` field set and the admin
has set baseline PSA level to `enforce` mode then the request will be
actively blocked and rejected. Existing pods with `.Host` probes
that are upgrading will not be impacted.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

No

###### What happens if we reenable the feature if it was previously rolled back?

N/A

###### Are there any tests for feature enablement/disablement?

N/A since there is no feature gate

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

> N/A

Author: I've updated this to include this info.. on second thoughts I have mentioned this in L246 which probably now needs a tweak? cause on upgrades all pods get recreated so if encountered with the PSA set they would fial?

###### What specific metrics should inform a rollback?

> N/A

Reviewer 4: Find the PSA denials metrics to help cluster-admins determine if they suddenly got a problem.

Author: ah you meant https://kubernetes.io/docs/concepts/security/pod-security-admission/#metrics ?
thanks for mentioning this.. learnt something new, didn't know these existed.
added them

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

N/A

### Monitoring Requirements

N/A

###### How can an operator determine if the feature is in use by workloads?

If pods have probes with `.Host` field set and PSA annotation is set on that pod's namespace
to a version where the new admission has been added, then it means the feature is enabled.

###### How can someone using this feature know that it is working for their instance?

Trying to create a pod with `.Host` field set in the probes will fail
like this:

> ```

Author: done

Error from server (Forbidden): error when creating "psa/fail-case-pod.yaml": pods "liveness-http-pass" is forbidden: violates PodSecurity "restricted:latest": probeHost (container "liveness" uses probeHost 135.45.63.4)
```

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
N/A

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

> N/A

Reviewer 4: PSA denials seem informative.

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

N/A

### Dependencies

None

###### Does this feature depend on any specific services running in the cluster?

No

### Scalability

N/A

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

No

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

No

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

N/A

###### What are other known failure modes?

N/A

> ###### What steps should be taken if SLOs are not being met to determine the problem?

Reviewer 4: these need to be filled in with at least references to the PSA docs about the difference between latest and pinned to versions.

Author: hmm sorry for the stupid q: "pinned to version" here is the one where this change will be introduced?
let me check a sample past PSA enhancement to see an example.

Author: ah I understand now, gotcha, included this info now in latest push

> # KEP-4940: Add PSA to block setting `.host` field from ProbeHandler and LifecycleHandler

Reviewer 5: Sorry for the silly suggestion:
```suggestion
# KEP-4940: Add Pod Security Admission to block setting `.host` field from ProbeHandler and LifecycleHandler
```

It took me a while to figure out what PSA was...

Author: not silly at all :) will change this.

Author: done

- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

We have a `Host` field that can be set from `TCPSocketAction` and
`HTTPGetAction` fields which are part of the `ProbeHandler` and
`LifecycleHandler` structs in Kubernetes. The `Host` field is used
for allowing users to specify another entity other than the
podIP (which is the default value) to which Kubelet should
perform probes to. However this opens it up for security attacks
since the `Host` field can be set to pretty much any value in the
system including security sensitive external hosts or localhost on
the node. Kubelet will be probing this set `Host` value which can

> lead to blind SSRF attacks.

Reviewer 6: The HTTP request `Host` is defaulted to the header value: https://github.com/kubernetes/kubernetes/blob/ff25881bc1fe9da1dda4ec82d294b92078ae2763/pkg/probe/http/request.go#L84

What does that control? Is it just used for setting the header value, or can it be used to route the request? Do we need to restrict the usage of a host header too?

Author: thanks @tallclair for taking a look..as far as I know its use for setting the header value and not for routing the request, but let me double check my understanding and circle back

### Goals

> * Add PSA to enable admins to restrict users from creating

Reviewer 5: ```suggestion
* Add Pod Security Admission (PSA) to enable admins to restrict users from creating
```

Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>

### Graduation Criteria

The PSA added will be done within a single release
and given there will be no feature gates for that,
there is no need for multi-release graduation criteria.
All related code will land within the same single release

#### Deprecation

> TBD

Author: add details for deprecating the .host field in lifecycle handler and probe handler here.

### Goals

* Add  Pod Security Admission (PSA) to enable admins to restrict
  users from creating probes with the `Host` field set.
* The Baseline Pod Security Standard (PSS) will be updated to enforce
  blocking this field so that it helps with easier adoption for
  workload operators given this is a known issue we want to prevent.

### Non-Goals

* Removing `.Host` field from the API and dropping support (It is
  unsaid rule that nothing can get removed from core Kubernetes API)

## Proposal

There is a long term plan to deprecate the existing TCP and HTTP probe
types in the API to replace them with ones with slightly different semantics.
See [KEP-4559](https://github.com/kubernetes/enhancements/pull/4558) for more
details. Given the unsolvable security problems with the Host field,
we do not plan to offer it in the new types.

Meanwhile, the older API is never going to go away. So we also want to
add PSA to allow admins to be able to restrict users from creating
probes with the Host field set when using the (about to be deprecated) API.
This is implemented by [kubernetes PR 125271](https://github.com/kubernetes/kubernetes/pull/125271)
that does exactly that.

> 

Author: oops I forgot to actually write the localhost part that @BenTheElder and @danwinship pointed out here: https://github.com/kubernetes/enhancements/pull/4942#discussion_r2122082241
will do a small FUP PR to fix that up

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

We have a `Host` field that can be set from `TCPSocketAction` and
`HTTPGetAction` fields which are part of the `ProbeHandler` and
`LifecycleHandler` structs in Kubernetes. The `Host` field is used
for allowing users to specify another entity other than the
podIP (which is the default value) to which Kubelet should
perform probes to. However this opens it up for security attacks
since the `Host` field can be set to pretty much any value in the
system including security sensitive external hosts or localhost on
the node. Kubelet will be probing this set `Host` value which can
lead to blind SSRF attacks.

## Motivation

**Probe-based Blind SSRF Attacks**: The current definition of TCP and
HTTP probes allows the user to specify an alternative hostname/IP to
connect to rather than the pod IP. (The expected use is for sending a
probe via a HostPort, NodePort, or LoadBalancer IP.) But this allows a
"blind SSRF" (Server-Side Request Forgery) attack, in which a pod can
trick kubelet into sending an HTTP GET request to an arbitrary URL (or

> portscanning TCP ports on arbitrary hosts). ([kubernetes #99425])

Reviewer 2: No need to repeat this; you just said all that in the summary.
But you should keep the Goals/Non-Goals section, even if there's only a single Goal.

Author: ack let me add back goalds/non-goals and remove this repeat paragraph

## Proposal

There is a long term plan to deprecate the existing TCP and HTTP probe
types in the API to replace them with ones with slightly different semantics.

> See the [KEP](https://github.com/kubernetes/enhancements/pull/4558) for more

Reviewer 2: ```suggestion
See [KEP-4559](https://github.com/kubernetes/enhancements/pull/4558) for more
```

Author: done.

details. Given the unsolvable security problems with the Host field,
we do not plan to offer it in the new types.

Meanwhile, the older API is never going to go away. So we also want to
add PSA to allow admins to be able to restrict users from creating
probes with the Host field set when using the (about to be deprecated) API.

> Here is the [draft PR](https://github.com/kubernetes/kubernetes/pull/125271)

Reviewer 2: (again I'd mention the PR number in the text. "This is implemented by [kubernetes #125271]" or something)

Author: done.

that does exactly that.

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> N/A

Reviewer 2: You can remove the HTML comments after you fill the section in.

Anyway, I think you say there's a risk, which is that some people may be depending on this functionality, and the mitigation for that is that the admin doesn't _have to_ block it. But there's no way to make the existing feature safe, so we think blocking it is the right thing to do.

## Design Details

Add a Baseline APILevel Pod Security Admission policy to allow admins of the
cluster to block users from setting `.host` field in:

* container.LivenessProbe.ProbeHandler.HTTPGet.Host
* container.ReadinessProbe.ProbeHandler.HTTPGet.Host
* container.StartupProbe.ProbeHandler.HTTPGet.Host
* container.LivenessProbe.ProbeHandler.TCPSocket.Host
* container.ReadinessProbe.ProbeHandler.TCPSocket.Host
* container.StartupProbe.ProbeHandler.TCPSocket.Host
* container.Lifecycle.PostStart.TCPSocket.Host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* container.Lifecycle.PreStop.TCPSocket.Host // Deprecated. TCPSocket is NOT supported as a LifecycleHandler and kept for backward compatibility.
* container.Lifecycle.PostStart.HTTPGet.Host
* container.Lifecycle.PreStop.HTTPGet.Host

### Test Plan

* Unit and E2E tests will be added to ensure the PSA works as expected

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.

> -->

Reviewer 2: None

Author: wasn't totally sure about this one... should we need any tests with probes that have that field set and then users upgrade and what happens?

Reviewer 2: "Prerequisite testing updates" is for the case where you're modifying something, and it turns out that the *existing* behavior is not very well tested, so you realize you're going to have to add tests for the existing behavior before it's possible to show that your *new* behavior works. That doesn't apply here.

Author: ah ack thanks for clarifying this.

##### Unit tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

> - `<package>`: `<date>` - `<test coverage>`

Reviewer 2: should fill this in based on the unit tests you'll be adding

Author: thanks! every time I try to click on https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit and scroll to the section the website crashes on me

##### Integration tests

<!--
Integration tests are contained in k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

> - <test>: <link to test coverage>

Reviewer 2: Should check how other parts of PSA are tested. If there are integration tests, you'll want to add one for this too. If not, then not.

Author: there is definitely some integration tests for PSA: https://github.com/kubernetes/kubernetes/blob/master/test/integration/auth/podsecurity_test.go generic ones

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

> - <test>: <link to test coverage>

Reviewer 2: likewise

Author: I don't see any https://github.com/kubernetes/kubernetes/tree/master/test/e2e/auth e2e's for pod security admission
in previous keps i see: https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/2579-psp-replacement#e2e-tests
unsure if that's acceptable or not
@deads2k should I be adding any new e2e's or we just go with integration tests as are today?

There is also these fixture tests: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/pod-security-admission/test which I know I will add I dunno if that's under e2e or just unit tests for testing the server?

Shouldn't I be adding tests like:
Test that pods with .host field set in probes are rejected when PSA is enabled with baseline level
Test that pods without .host field set in probes are allowed when PSA is enabled with baseline level
Test that existing pods with .host field set continue to work when PSA is enabled
Test that pods with .host field set are allowed when PSA is disabled or using an older version

what category are these in? e2e? integration?

> ### Graduation Criteria

Reviewer 2: need some explanation here, even if it's not Alpha/Beta/GA

Author: added.

- Address feedback on usage/changed behavior, provided on GitHub issues
- Deprecate the flag
-->

### Upgrade / Downgrade Strategy

Any older pods with this field set should not be affected
with the above solution. Only newer pods getting created
with the field will be alerted.

Users who are using this field can switch to using exec
probes moving forward which should unblock them given exec
probes can provide the same functionality.

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and nodes?
- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

> N/A

Author: since its only across a single component

## Production Readiness Review Questionnaire

<!--

Production readiness reviews are intended to ensure that features merging into
Kubernetes are observable, scalable and supportable; can be safely operated in
production environments, and can be disabled or rolled back in the event they
cause increased failures in production. See more in the PRR KEP at
https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.

The production readiness review questionnaire must be completed and approved
for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

> ### Feature Enablement and Rollback

Reviewer 2: So when the PRR talks about "enablement" it doesn't mean "using the feature", it means "configuring the cluster in a way that users _can_ use the feature". If the change to the Pod Security Standards is active by default, then "enablement" is exactly the same as "upgrading"; the new feature is enabled when you upgrade to the release that contains it.

Some things to consider/discuss here are:
- There's no effect on clusters that aren't using PSA, and no effect on clusters that aren't using Host probes.
- If people are using, eg, `pod-security.kubernetes.io/enforce-version: v1.31`, then there will be no effect even if they are using Host probes, because they're explicitly requesting an older version of the standards.

Author: ack thanks @danwinship for outlining this. will update this part with the relevant information

Reviewer 3: I believe this is resolved now.

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->
N/A

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

<!--
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->
N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

<!--
Even if applying deprecation policies, they may still surprise some users.
-->
N/A

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->
N/A

###### How can an operator determine if the feature is in use by workloads?

<!--
Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
checking if there are objects with field X set) may be a last resort. Avoid
logs or events for this purpose.
-->

> TBD

Reviewer 2: looking at the PSA annotations on namespaces

Author: done.

###### How can someone using this feature know that it is working for their instance?

<!--
For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
for each individual pod.
Pick one more of these and delete the rest.
Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
and operation of this feature.
Recall that end users cannot usually observe component logs or access metrics.
-->

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [ ] Other (treat as last resort)
  - Details:

> TBD

Reviewer 2: Trying to create a pod with Host probes will fail/warn/audit.

as node-level agents (e.g. specific version of CRI). Focus on external or
optional services that are needed. For example, if this feature depends on
a cloud provider API, or upon an external software-defined storage or network
control plane.

For each of these, fill in the following—thinking about running existing user workloads
and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->

N/A

### Scalability

N/A

###### Will enabling / using this feature result in any new API calls?

<!--
Describe them, providing:
  - API call type (e.g. PATCH pods)
  - estimated throughput
  - originating component(s) (e.g. Kubelet, Feature-X-controller)
Focusing mostly on:
  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->

> TBD

Reviewer 2: No

###### Will enabling / using this feature result in introducing new API types?

<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

<!--
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
-->

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

<!--
Look at the [existing SLIs/SLOs].

Think about adding additional work or introducing new steps in between
(e.g. need to do X to start a container), etc. Please describe the details.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

> TBD

Reviewer 2: No

[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
-->
No.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

<!--
Focus not just on happy cases, but primarily on more pathological cases
(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
If any of the resources can be exhausted, how this is mitigated with the existing limits
(e.g. pods per node) or new limits added by this KEP?

Are there any tests that were run/should be run to understand performance characteristics better
and validate the declared limits?
-->

No.

### Troubleshooting

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

###### How does this feature react if the API server and/or etcd is unavailable?

> TBD

Reviewer 2: N/A; it's an apiserver feature
